### DISCUSSION TITLE
Help with MT7668 Driver kernel source

### DISCUSSION THREAD

**Discussion by mryti** [devmfc/debian-on-amlogic#15] [General]
Hi, Mate.
I have a Amlogic 905x2 TV Box (ZTE B860H V5) with MT7668 sdio WiFi combo chip. I successfully booted this device with armbian and openwrt, but the WiFi and Bluetooth isn't working at all. I tried to compile the driver from available source in GitHub [this repo](https://github.com/yHpgi/MT7668). it's successfully compiled, but when i try to load the driver, suddenly I got kernel panic.
It will be nice if you can share the kernel source with the MT7668 driver. I would love to do trials and errors in the future to get (at least) the WiFi work in linux. 
I am blind rn, any help will be really appreciated. Thank you.

**Comment by devmfc** [comment]
hi,
I will put it on my list to clean that code a bit and put it in a Github repo. I already started with that, but it takes some more time. My problem is that this project is pure a hobby project in my scarce spare time. But because it is just a hobby, just to tinker/hack around, the quality is not always what I do professionally. That is also the reason I don't just want to publish it as it is. 

Also the link you provided links also to the github of ChalesYu. This is one of the sources I use for one or two other wifi drivers and he is clearly more into wifi than I am ;-).  His code looks a lot like my code, but might be better. So maybe I will rebase my necessary patches to that one.

Anyway, the problem you encountered is probably causes by a bug in the g12a sdio controller hardware. So you could try to reduce the MTU size of wlan0 to 1200 bytes or so. Maybe it will just work in that case.  And otherwise, just wait for me while I put everything in a repository.

**Reply by mryti** [reply]
I appreciate you reaching out. now I can see a bright spot ~
Anyway, thank you for your excellent work!.

**Comment by mryti** [comment]
Hi, I want to say Thank you!
I successfully load the driver from your source. Tested on Armbian and OpenWrt.

But, wifi unavailable after reboot. here some logs that i got:
```
[   14.600434] DEVMFC: mtk_sdio_probe
[   14.602564] wlan mmc2:0001:1: Mediatek MT7668 wlanProbe
[   14.606014] wlan mmc2:0001:1: DEVMFC: g12a workaround, setting mtu to 1408.
[   15.123915] DEVMFC: halRxWaitResponse packet longer than buf u4PktLen[57]
[   15.127174] [wlan][539]wlanAccessRegisterStatus:(INIT ERROR) wlanAccessRegisterStatus: incorrect ucEID. ucSetQuery = 0x0
[   15.137955] [wlan][539]wlanAccessRegister:(INIT ERROR) Event reports address incorrect. 0x80021000, 0x00000000.
[   15.148003] [wlan][539]wlanSetChipEcoInfo:(INIT ERROR) wlanSetChipEcoInfo >> get TOP_HVR failed.
[   15.157201] [wlan][539]wlanProbe:(INIT ERROR) probe failed -5
[   15.163014] wlan: probe of mmc2:0001:1 failed with error -1
```
![image](https://github.com/devmfc/debian-on-amlogic/assets/91925593/e72d9201-acb8-4ffc-894c-74a60d498876)
wifi will back to work after I unplug the power plug and plug it back in.

And one more thing, >.<
I can't use the wifi as access point (OpenWrt, untested on Armbian)
the wifi can not be turned on when I set it on access point mode, but work fine on station/client.
is it possible to use this mt7668 as an access point?
 
Best Regards!

**Reply by devmfc** [reply]
You're welcome.
I think it doesn't load because the bluetooth drivers interfere in some way. I think you should blacklist the Mediatek bluetooth drivers. I never got bluetooth working for this wifi module.

Whether AP mode works: I don't know. Never tried that with this wifi module.

I published the source code, I will clean it up a little more, but that is all I can do I'm afraid. I'm not a wifi expert, in fact, I hate wifi except for my phone and tablet.

**Reply by mryti** [reply]
I got you. Anyway, Thank you.

**Reply by worstcase-scenario** [reply]
Perhaps this info can help you:

I also have the Mt-7668-chip on my android box. I am running EmuElec on it and the WiFi works, but not the bluetooth. Then I found this on the net: https://ps4linux.com/mt7668-wifi-bluetooth-drivers-linux-compile/
and contacted the developer of EmuELEC, who was able to implement the driver successfully. Now both WiFi AND Bluetooth is working.

**Reply by devmfc** [reply]
Thanks for the link! I will certainly take a look at it, but EmuElec is AFAIK based on Coreelec, which in turn is based on an older vendor kernel. So if it works there will not mean it works on mainline. 

But every bit of info is welcome, thanx!

**Reply by Pro-me3us** [reply]
Do you know if the mt7668 BT driver's BLE functionality is working with either the Khadas mt7668 WiFi/BT combo drivers, or the mainline Linux mt7668 BT driver?

I've tried both and could only get BR/EDR functionality working.  I don't know if this is because BLE functionality is missing from the Linux drivers or if it's just my compiled version?

When you run `hciconfig hci0 commands` do any of the LE commands show up? Or running a `hcitool lescan`, do any BLE devices show up?

**Reply by devmfc** [reply]
Really? With mainline Linux? In that case you have had more success than I have. That's great news.
I never got the Bluetooth part working at all.

Can you please tell me which image you used on which box and which kernel version, which firmware version?  Maybe I can make it work as well and proceed from there?

**Reply by Pro-me3us** [reply]
I'm running CoreELEC and EmuELEC, on Linux kernels 4.9 and 5.15, on an Amlogic g12b (s922x).  CE-NG (4.9) uses the [Google Coral mt7668 mod](https://coral.googlesource.com/mt7668-bluetooth-mod/) (bases of what Khadas posted).  The CE version is available [here](https://github.com/CoreELEC/MT7668/).  It also has conditions for newer kernels, but I've only tested it on CE-NO (5.15), and it works.

[EmuELEC](https://github.com/shantigilbert/mt7668-wifi-bt) (also kernel 4.9) uses the mt7668 driver taken from the [Fire TV 2nd gen Cube](https://www.amazon.com/gp/help/customer/display.html?nodeId=201452680) with a small modification taken from the Google Coral source. Amazon has a few devices that use the m7668 chip including 2nd gen Cube, 4K Stick (2018), and [Echo Studio](https://www.amazon.com/gp/help/customer/display.html?nodeId=201626480).  

In CE-NO (5.15) the [mainline Linux drivers](https://github.com/CoreELEC/linux-amlogic/blob/5.15.137_202409/drivers/bluetooth/) are included as well.  I'm not sure exactly which mainline linux version is being used, they don't match up exactly to [these](https://github.com/torvalds/linux/tree/v5.15/drivers/bluetooth), but is very close.

I did have to edit [btmtksdio.c](https://github.com/CoreELEC/linux-amlogic/blob/5.15.137_202409/drivers/bluetooth/btmtksdio.c) 

[Deleted line 384](https://github.com/CoreELEC/linux-amlogic/blob/c375891357fced5c5a8270ebb899fc69b082187c/drivers/bluetooth/btmtksdio.c#L334C1-L334C20)
`u8 evt = hdr->evt;`

[Modified line 359 to](https://github.com/CoreELEC/linux-amlogic/blob/c375891357fced5c5a8270ebb899fc69b082187c/drivers/bluetooth/btmtksdio.c#L359C2-L359C29)
`if (hdr->evt == HCI_EV_VENDOR) {`

With all mt7668 BT versions on both kernel 4.9 and 5.15, BR/EDR worked but BLE failed to detect any devices during a scan.  The LE commands are absent when checking with `hciconfig hci0 commands`. I don't know if this is because of a driver/device problem or the LE functionality is missing from the drivers.  I know that the s905x3 running EmuELEC (4.9) exhibited the same BR/EDR functionality without BLE.

The mainline Linux driver is slightly better than the Google Coral/Amazon version, in that it allowed pin# authentication without failing.

A few variations of the PS4 also use the mt7668 chip.  I was told that bluetooth is working on kernel 6.1, but I don't have any more details, such as if that is using the mainline or google coral source.
